src/hotspot/share/runtime/handshake.cpp

```diff
@@ -370,6 +375,10 @@ void Handshake::execute(HandshakeClosure* hs_cl, ThreadsListHandle* tlh, JavaThr
     return;
   }
 
+  if (UseSystemMemoryBarrier) {
+    SystemMemoryBarrier::emit();
+  }
```

Comment:
It's not clear to me why this emit() is here.

Reply:
add_operation() arms poll.
try_process() reads JavaThread state.

Since are allowing the CPU to do reordering of instructions, try_process() can see e.g. state in_native after arming the poll.
Normally this would be safe, but since we now allow reordering it's not.
By emitting the system membar we know that if we see in_native that thread will see the armed poll.

Long comment from me above, but yes it's not clear that add_operation() and try_process() are effected by this race when using system membar.

Reply:
How about a comment above the emit():
// Separate the poll() in add_operation() above() from the read of JavaThread state
// in the try_process() call below.

Reply:
Fixed